### PR TITLE
[s11n 2/7] Make session serialisation functions public

### PR DESCRIPTION
**Pull request by mpg**
## Description

This is a side-port to development of PR https://github.com/ARMmbed/mbedtls-restricted/pull/568 that has already been merged to baremetal.

**Dependencies:** based on #2764 - see "Background" below.

The goal of this PR is to make the existing functions for serializing a `ssl_session` structure public. In the process, their API is improved and a non-zero amount of (unit-)testing is added.

## Background

This is the second in a series of seven PRs implementing serialization of a full SSL context (not just the session), under some restrictions.

The PRs in the series are:
 
1. [r563](https://github.com/ARMmbed/mbedtls-restricted/pull/563) -> #2764 Refactor derive keys
2. [r568](https://github.com/ARMmbed/mbedtls-restricted/pull/568) -> #2766 Make session serialisation functions public __(this PR)__
3. [r570](https://github.com/ARMmbed/mbedtls-restricted/pull/570) -> #2785 Make session format ABI-independent
4. [r567](https://github.com/ARMmbed/mbedtls-restricted/pull/567) -> #2788 Add version + config check to SSL tickets
5. [r572](https://github.com/ARMmbed/mbedtls-restricted/pull/572) -> #2789 Context serialisation: API and documentation
6. [r575](https://github.com/ARMmbed/mbedtls-restricted/pull/575) -> #2791 Serialization test
7. [r616](https://github.com/ARMmbed/mbedtls-restricted/pull/616) -> #2792 Implement context serialization

The full series is already merged in baremetal. It needs upstreaming for the benefit of the Could team, who will be the main user, and relies on upstream releases.

## Side-porting details

Side-porting was done with `git cherry-pick`; most commit were picked with no conflicts. Out of the remaining ones, most of the conflicts were trivial (only the context had changed due to independent addition at the same place). Only a few commits required thoughtful conflict resolution.

All non-trivial conflicts are related to the fact the the `config.h` option `MBEDTLS_SSL_KEEP_PEER_CERTIFICATE` was introduced prior to this PR in development, while it was introduced after it in baremetal.

The table shows how each side-ported commit was handled:

- (nothing) cherry-picking worked without conflict
- "t" only trivial context-related conflicts (eg. independent additions to the same place)
- "r" manual rework after cherry-picking
- "R" the commit was reconstructed manually (moving code around)
- "+" added in this side-port PR

| resolution | original commit | commit message |
|---| ---          | --- |
|   | 2843fe10b9aa | Declare and document session save/load functions |
| R | 91f4ca2ed144 | Move session save/load function to ssl_tls.c |
| r | ef4ae611e4ca | Add support for serialisation session with ticket |
|   | fbb44a422f20 | Save session in serialised form in ssl_client2. |
|   | 37a5324c747f | Add mbedtls_ssl_get_session_pointer() |
|   | 57a348ba8cf0 | Add tests for session copy without serialisation |
| r | 32ce596c35b2 | Improve save API by always updating olen |
| t | dfa5a7ae7610 | Start adding unit test for session serialisation |
| t | 16f6bb1aa379 | Improve load-save test with tickets and certs |
| t | 1ba5c68503bd | Disable test for load-save identity |
|   | 98fccc3f6ab9 | Add test for session_save() on small buffers |
| r | 5709811dd2e2 | Add test for session_load() from small buffers |
|   | aab6204dc133 | Fix populate_session() and its usage in tests |
| t | 2a62a05688a7 | Add test that save-load is the identity |
|   | 5363e1f496f1 | Add list of coupled functions to struct definition |
|   | 081b15231f20 | Fix another wrong check for errors in test code |
|   | d1a5451fb5ab | Fix style issues and typos in test code |
|   | 749312fb8a19 | Fix undeclared dependency on FS_IO in test code |
|   | 2d8847e84dc6 | Add a ChangeLog entry for session serialisation |
|   | 4bb1b99c7ff9 | Demonstrate safe usage (zeroize) in ssl_client2 |
|   | e0cd1d018436 | Improve documentation |
| t | 35ccdbb636af | Normalize spelling to serialiZation |
| + | (new) | Fix bug in cert digest serialisation |
| + | (new) | Fix serialization tests for !SSL_KEEP_PEER_CERT |

The added commits at the end fix bugs that were introduced in the side-porting (related to different order of PRs in development compared to baremetal) but not present in baremetal.

## Review notes

Reviewers are strongly encouraged to review commit by commit, as the final diff is unlikely to make sense. Please not that the first commits (up to "Use more specific name in debug message for testing") are part of #2764 and don't need to be re-reviewed here (except checking that they're indeed part of #2764).

### CODE REVIEW

**Comment by mpg** [comment]
This now passes CI, removing "needs work" and adding "needs review".

**Comment by jarlamsa** [comment]
#2764 history was rewritten, rebased over that.

**Comment by jarlamsa** [comment]
Rebased on top of #2764 and resolving a trivial conflict in ChangeLog.

File: ChangeLog [hunk 1]
```diff
@@ -139,6 +142,15 @@ Changes
...
      Contributed by Philippe Antoine (Catena cyber).
 
+API Changes
+   * Extend the MBEDTLS_SSL_EXPORT_KEYS to export the handshake randbytes,
+     and the used tls-prf.
+   * Add public API for tls-prf function, according to requested enum.
+   * Add DER-encoded test CRTs to library/certs.c, allowing
+     the example programs ssl_server2 and ssl_client2 to be run
+     if MBEDTLS_FS_IO and MBEDTLS_PEM_PARSE_C are unset. Fixes #2254.
+   * The HAVEGE state type now uses uint32_t elements instead of int.
```
Comment (Patater) [inline ChangeLog]: Looks like these were added to an older version, not the most recent version.

Comment (jarlamsa) [reply ChangeLog]: Will be fixed in the 7th PR in the series.

**Comment by jarlamsa** [comment]
Rebased

**Comment by jarlamsa** [comment]
@hanno-arm please re-review
@Patater the comment has been adressed in #2792 please re-review.

File: programs/ssl/ssl_client2.c [hunk 2]
```diff
@@ -38,6 +38,8 @@
 #define mbedtls_calloc     calloc
 #define mbedtls_free       free
 #define mbedtls_exit            exit
+#define mbedtls_calloc      calloc
+#define mbedtls_free        free
```
Comment (Patater) [inline programs/ssl/ssl_client2.c]: `mbedtls_calloc` and `mbedtls_free` are already defined in this file. Why define them again? (Feel free to fix in PR 7/7)

**Comment by jarlamsa** [comment]
Rebased without conflicts

**Comment by jarlamsa** [comment]
@hanno-arm @Patater please re-review

**Comment by jarlamsa** [comment]
All of the serialisation PRs now have 2 approving reviews and passing CI. Thus considering all of them ready for merge.

**Pull request merged**